pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,261 @@
+}
+
+// PreFilter validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`.
+// If so, the scheduling process will be interrupted directly to avoid the partial Pods holding system resources
+// until a timeout. It will reduce the overall scheduling time for the whole group
+func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, p *v1.Pod) *framework.Status {
+	podGroupName, minAvailable := GetPodGroupLabels(p)
+	if podGroupName == "" || minAvailable <= 1 {
+		return framework.NewStatus(framework.Success, "")
+	}
+
+	total := cs.calculateTotalPods(podGroupName, p.Namespace)
+	if total < minAvailable {
+		klog.V(3).Infof("The count of podGroup %v/%v is not up to minAvailable(%d) in PreFilter: %d",
+			p.Namespace, podGroupName, minAvailable, total)
+		return framework.NewStatus(framework.Unschedulable, "less than minAvailable")
+	}
+
+	return framework.NewStatus(framework.Success, "")
+}
+
+func (cs *Coscheduling) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Permit is the functions invoked by the framework at "permit" extension point.
+func (cs *Coscheduling) Permit(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
+	podGroupName, minAvailable := GetPodGroupLabels(p)
+	if podGroupName == "" || minAvailable <= 1 {
+		return framework.NewStatus(framework.Success, ""), 0
+	}
+
+	namespace := p.Namespace
+	running := cs.calculateRunningPods(podGroupName, namespace)
+	waiting := cs.calculateWaitingPods(podGroupName, namespace)
+	current := running + waiting + 1
+
+	if current < minAvailable {
+		klog.V(3).Infof("The count of podGroup %v/%v is not up to minAvailable(%d) in Permit: running(%d), waiting(%d)",
+			namespace, podGroupName, minAvailable, running, waiting)
+		// TODO Change the timeout to dynamic value depends on the size of the `PodGroup`
+		return framework.NewStatus(framework.Wait, ""), 10 * PermitWaitingTime
+	}
+
+	klog.V(3).Infof("The count of podGroup %v is up to minAvailable(%d) in Permit: running(%d), waiting(%d)",
+		podGroupName, minAvailable, running, waiting)
+	cs.FrameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		if waitingPod.GetPod().Namespace == namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
+			klog.V(3).Infof("Permit allows the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
+			waitingPod.Allow(cs.Name())
+		}
+	})
+
+	return framework.NewStatus(framework.Success, ""), 0
+}
+
+// Unreserve is called by the scheduling framework when a reserved pod was
+// rejected in a later phase.
+func (cs *Coscheduling) Unreserve(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) {
+	podGroupName, exist := p.Labels[PodGroupName]
+	if !exist {
+		return
+	}
+
+	cs.FrameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
```

Comment:
@Huang-Wei do you think it's possible that a rejected pod is called more than once? I feel like it is, due to the asynchronous implementation of Reject.

Reply:
@alculquicondor Yes. It can be called multiple times, and the first call with `Unschedulable` status will be caught by the internal channel; other calls afterwards will be treated as no-ops.

Reply:
sg